WIP: Support multiple PVCs in default JH UI, add example 3rd-party UI

ioandr opened the pull request:
This is a PR that:
* Adds support for handling multiple PVCs via the default JupyterHub Spawner UI
* Adds a 3rd-party UI for Arrikto Rok as an example of how to extend the default UI and integrate with alternative storage platforms.

This PR depends on PR kubeflow/kubeflow#1839.

The main changes of this PR are:
* It extends the default Spawner UI with two new Spawner form fields which allow users to declare which volumes will be attached to their notebook. Users can configure:
  * a single “Workspace” Volume (mounted at the home directory)
  * multiple “Data” Volumes
* It adds support for attaching existing and provisioning new Volumes to a Jupyter notebook. For existing volumes, the UI autocompletes volume-related fields, e.g., `size`.
* It obsoletes the existing `disks` and `notebookPVCMount` ksonnet parameters, along with their assignments in `jupyterhub_config.py`. Their functionality is covered by the `<uidir>/config.yaml` mechanism added in PR kubeflow/kubeflow#1839, more on this below.
* We also introduce an explicit `storageClass` ksonnet parameter, which defaults to `"null"`,
so it fall backs to the current behavior.

Making these changes solves a few standing issues:
* The user can easily attach newly created or existing Volumes to his Jupyter Notebook, via the default Spawner UI: This closes kubeflow/kubeflow#541, kubeflow/kubeflow#34. 
* The admin can easily configure shared Volumes for all Notebooks, by specifying the `workspaceVolume` and `dataVolumes` fields in `config.yaml`

Known issues / next steps: [@jlewi, please comment :) ]
* We still need a way to document the format of `config.yaml`. For now, please see the contents of the included in `kubeflow/core/ui/default/config.yaml` and the code in `kubeflow/core/ui/default/spawner.py` on the supported YAML attributes. Where would be the best place to document this explicitly, as the supported format for `<uidir>/config.yaml`?
* Similarly, where would be the best place to document migrating away from the older `disks`, `notebookPVCMount` parameters?

kkasravi commented:
> We still need a way to document the format of config.yaml. For now, please see the contents of the included in kubeflow/core/ui/default/config.yaml and the code in kubeflow/core/ui/default/spawner.py on the supported YAML attributes. Where would be the best place to document this explicitly, as the supported format for <uidir>/config.yaml?

You could define a json schema so that it could be checked/validated

jlewi commented:
Does this PR need to be sync'd and updated now that #1839 has been merged? Also I'm guessing this PR needs to be updated now that Jupyter is in its own package and not core.

ioandr commented:
> Does this PR need to be sync'd and updated now that #1839 has been merged? Also I'm guessing this PR needs to be updated now that Jupyter is in its own package and not core.

Jeremy, thanks for the heads-up. We are already working on the rebase you mentioned to make all necessary updates to remove the WIP label from this PR.

ioandr commented on kubeflow/core/jupyterhub.libsonnet:
```diff
@@ -224,13 +235,15 @@
           resources: [
             "pods",
             "persistentvolumeclaims",
+            "secrets",
```
Hi Kam. To our knowledge, gcp_scecret_name gets mounted to the user's notebook pod as a separate volume and was included in a previous version of `jupyterhub.libsonnet`. 
Our change here does not deal with mounting secrets to the notebook pod; instead, it is about giving the JupyterHub pod the right privileges to create Secrets via the K8s API. We believe this is essential to currently support 3rd party UIs and emit Notebook/Profile CRDs in the near future.

ioandr commented on kubeflow/core/jupyterhub_config.py:
```diff
@@ -0,0 +1,114 @@
+###################################################
+c.JupyterHub.spawner_class = spawner.KubeFormSpawner
+
+c.KubeSpawner.cmd = 'start-singleuser.sh'
```
We are inline with the idea of emitting Notebook/Profile CRDs in the near future, however we need to point out that this PR still relies on KubeSpawner and emits Pods at the very end. We are in the direction of directly mapping form UI fields to Notebook attributes in a future PR, when #1855 gets merged.

ioandr commented:
> > Does this PR need to be sync'd and updated now that #1839 has been merged? Also I'm guessing this PR needs to be updated now that Jupyter is in its own package and not core.
> 
> Jeremy, thanks for the heads-up. We are already working on the rebase you mentioned to make all necessary updates to remove the WIP label from this PR.

Hi again Jeremy, we have updated our branch and removed the WIP label from this PR. Also, tests are passing in the CI cluster.

ioandr commented:
/retest

jlewi commented:
@pdmack @kkasravi anyone want to volunteer to take the lead on reviewing this?

kkasravi commented:
I can pick it up since it's closely intertwined with notebooks

ioandr commented:
> I can pick it up since it's closely intertwined with notebooks

Hi Kam,

we acknowledge that this PR is quite big and comes with many code changes that might be time-consuming to review. As such, we are on standby to elaborate on our implementation and discuss individual parts, if this is of some help to you.

Thank you for your time on this.

kkasravi commented on kubeflow/jupyter/ui/default/template.html:
```diff
@@ -1,110 +1,233 @@
+              </div>
+              <div class="col-sm-2" style="width: 14%">
+                <select class="form-control" id="ws_access_modes" name="ws_access_modes" required>
+                  <option value="ReadWriteOnce">RWO</option>
```
should there be a tooltip or alt text for these?

ioandr replied:
Hi Kam,

we have already added a tooltip for the Access Mode `<select>` elements that shows up when the Type of the Volume is set to 'Existing`, to inform about autocompletion. As such, instead of a second tooltip, we suggest that the available options of  the Access Mode dropwdown menu should be updated as follows :
- RWO --> ReadWriteOnce
- RWX --> ReadWriteMany
- ROX --> ReadOnlyMany

kkasravi commented:
/lgtm

jlewi commented:
@kkasravi @ioandr How's this PR coming?

jlewi commented:
Looks there was some discussion in slack about this PR? The gist of which was this PR is ready to go but it looks like there might have been some changes since the last LGTM.

jlewi commented:
/approve

Looks like jupyter needs an OWNERs file.

ioandr commented:
/retest

vkoukis commented:
/retest
